src/Gui/DlgUnitsCalculatorImp.cpp

```diff
@@ -95,31 +96,42 @@ void DlgUnitsCalculator::reject()
     QDialog::reject();
 }
 
-void DlgUnitsCalculator::unitValueChanged(const Base::Quantity& unit)
+void DlgUnitsCalculator::textChanged(QString unit)
 {
-    actUnit = unit;
-    valueChanged(actValue);
+	valueChanged(actValue);
 }
 
 void DlgUnitsCalculator::valueChanged(const Base::Quantity& quant)
 {
-    if (actUnit.isValid()) {
-        if (actUnit.getUnit() != quant.getUnit()) {
-            ui->ValueOutput->setText(tr("Unit mismatch"));
-            ui->pushButton_Copy->setEnabled(false);
-        } else {
-            double value = quant.getValue()/actUnit.getValue();
-            QString val = QLocale::system().toString(value, 'f', Base::UnitsApi::getDecimals());
-            QString out = QString::fromLatin1("%1 %2").arg(val, ui->UnitInput->text());
-            ui->ValueOutput->setText(out);
-            ui->pushButton_Copy->setEnabled(true);
-        }
-    } else {
-        //ui->ValueOutput->setValue(quant);
-        ui->ValueOutput->setText(quant.getUserString());
-        ui->pushButton_Copy->setEnabled(true);
-    }
-
+	// first check the unit, if it is invalid, getTypeString() outputs an empty string 
+    if (Base::Unit(ui->UnitInput->text()).getTypeString().isEmpty()) {
+		ui->ValueOutput->setText(tr("unknown unit: ") + ui->UnitInput->text());
+		ui->pushButton_Copy->setEnabled(false);
+	} else { // the unit is valid
+		// we can only convert units of the same type, thus check
+		if (Base::Unit(ui->UnitInput->text()).getTypeString() != quant.getUnit().getTypeString()) {
+			ui->ValueOutput->setText(tr("unit mismatch"));
+			ui->pushButton_Copy->setEnabled(false);
+		} else { // the unit is valid and has the same type
+			double convertValue = Base::Quantity::parse(QString::fromLatin1("1") + ui->UnitInput->text()).getValue();
+			// we got now e.g. for "1 in" the value '25.4' because 1 in = 25.4 mm
```

Comment:
Please use spaces instead of tabs.  
Also I propose swapping `we got now` to `we have now for example, for "1 in." the value '25.4' because 1 in. = 25.4mm

Reply:
Inch is abbreviated `in`, without the dot at the end. So not `in.`